File: src/main/scala/nl/rug/jbi/search/ssap/model/Pattern.scala
package nl.rug.jbi.search.ssap.model

import scala.collection.mutable
import scala.collection.mutable.ArrayBuffer

class Pattern(val name: String, val instances: mutable.Buffer[Instance]) {

  def toXml() = {
    <pattern name={name}>
      {instances.map(_.toXml())}
    </pattern>
  }

  override def equals(o: Any) = o match {
    case that: Pattern => that.name == name
    case _ => false
  }
  override def hashCode = name.hashCode()
}

object Pattern {
  def fromXml(node: scala.xml.Node):Pattern = {
    val p:Pattern = new Pattern((node \ "@name").text, ArrayBuffer[Instance]())
    Instance.resetId()
    (node \\ "instance").map(i => p.instances+=Instance.fromXml(i))
    p
  }

  def getGroup(name:String):String = name match {
    case "Factory Method" => "Creational"
    case "Prototype" => "Creational"
    case "Singleton" => "Creational"
    case "(Object)Adapter-Command" => "Structural-Behavioral"
    case "Composite" => "Structural"
    case "Decorator" => "Structural"
    case "Observer"  => "Behavioral"
    case "State-Strategy"  => "Behavioral"
    case "Template Method" => "Behavioral"
    case "Visitor"   => "Behavioral"
    case "Proxy"     => "Structural"
    case "Proxy2"    => "Structural"
    case _ => ""
  }
}

File: ssap/src/main/java/nl/rug/jbi/search/ssap/model/Pattern.java
package nl.rug.jbi.search.ssap.model;

import com.thoughtworks.xstream.annotations.XStreamImplicit;
import nl.rug.jbi.search.ssap.util.Constants;

import java.util.List;
import java.util.Objects;

public class Pattern {
    public final String name;
    @XStreamImplicit
    public List<Instance> instanceList;

    private Pattern (String name, List<Instance> instanceList) {
        this.name = name;
        this.instanceList = instanceList;
    }

    public String getName () {
        return this.name;
    }

    public String getGroup (String name) {
        switch (name) {
            case Constants.FACTORY_METHOD:
            case Constants.SINGLETON:
            case Constants.PROTOTYPE:
                return Constants.CREATIONAL;
            case Constants.ADAPTER_COMMAND: return Constants.STRUCTURAL_BEHAVIORAL;
            case Constants.COMPOSITE:
            case Constants.DECORATOR:
            case Constants.PROXY:
            case Constants.PROXY2:
                return Constants.STRUCTURAL;
            case Constants.OBSERVER:
            case Constants.STATE_STRATEGY:
            case Constants.TEMPLATE_METHOD:
            case Constants.VISITOR:
                return Constants.BEHAVIORAL;
            default: return "";
        }
    }

    @Override
    public boolean equals (Object obj) {
        if (Pattern.class == obj.getClass()) {
            return Objects.equals(((Pattern) obj).name, name);
        }
        return false;
    }

    @Override
    public int hashCode () {
        return name.hashCode();
    }
}
